src/reference-manual/compiler-optimizations.Rmd

```diff
@@ -0,0 +1,487 @@
+
+Example Stan program:
+
+```
+data {
+  real y;
+}
+model {
+  real x = y + 1;
+}
+```
+
+Compiler representation of program **before AD-level optimization** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+input_vars {
+  real y;
+}
+
+log_prob {
+  real x = (y + 1);
+}
+```
+
+Compiler representation of program **after AD-level optimization** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+input_vars {
+  real y;
+}
+
+log_prob {
+  data real x = (y + 1);
+}
+```
+
+#### Partial evaluation
+
+Partial evaluation searches for expressions that can be replaced with a faster, simpler, more memory efficient, or more numerically stable expression that has the same meaning.
+
+Example Stan program:
+
+```
+model {
+  real a = 1 + 1;
+  real b = log(1 - a);
+  real c = a + b * 5;
+}
+```
+
+Compiler representation of program **before partial evaluation** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+log_prob {
+  real a = (1 + 1);
+  real b = log((1 - a));
+  real c = (a + (b * 5));
+}
+```
+
+Compiler representation of program **after partial evaluation** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+log_prob {
+  real a = 2;
+  real b = log1m(a);
+  real c = fma(b, 5, a);
+}
+```
+
+
+
+### **Oexperimental** Optimizations
+
+
+#### One step loop unrolling
+
+One step loop unrolling is similar to [static loop unrolling](#orge4599d2), but it only 'unrolls' the first iteration of a loop, and can therefore work even when the total number of iterations is not predictable. This can speed up a program by providing more opportunities for further optimizations such as partial evaluation and lazy code motion.
+
+Example Stan program:
+
+```
+data {
+  int n;
+}
+transformed data {
+  int x = 0;
+  for (i in 1:n) {
+    x += i;
+  }
+}
+```
+
+Compiler representation of program **before one step static loop unrolling** (simplified from the output of `--debug-transformed-mir-pretty`):
+
+```
+prepare_data {
+  data int n = FnReadData__("n")[1];
+  data int x = 0;
+  for(i in 1:n) {
+    x = (x + i);
+  }
+}
+```
+
+Compiler representation of program **after one step static loop unrolling** (simplified from the output of `--debug-optimized-mir-pretty`):
+
+```
+prepare_data {
+  data int n = FnReadData__("n")[1];
+  int x = 0;
+  if((n >= 1)) {
+    x = (x + 1);
+    for(i in (1 + 1):n) {
+      x = (x + i);
+    }
+  }
+}
+```
+
+#### Expression propagation
+
+<a id="org4e2cadd"></a> Constant propagation replaces uses of a variable which is known to have a constant value `E` with that constant `E`. This often results in recalculation of the expression, but provides more opportunities for further optimizations such as partial evaluation. Expression propagation is always followed by [lazy code motion](#org0c605fa) to avoid unnecessarily recomputing expressions.
```

Comment:
```suggestion
<a id="org4e2cadd"></a> Constant propagation replaces the uses of a variable known to have a constant value `E` with that constant `E`. This often results in recalculating the expression but provides more opportunities for further optimizations such as partial evaluation. Expression propagation is always followed by [lazy code motion](#org0c605fa) to avoid unnecessarily recomputing expressions.
```